rootfs/etc/nginx/lua/certificate.lua

```diff
@@ -0,0 +1,50 @@
+local ssl = require("ngx.ssl")
+local configuration = require("configuration")
+
+local _M = {}
+
+local function set_pem_cert_key(pem_cert_key)
+    local der_cert, der_cert_err = ssl.cert_pem_to_der(pem_cert_key)
+    if not der_cert then
+        return "failed to convert certificate chain from PEM to DER: " .. der_cert_err
+    end
+
+    local set_cert_ok, set_cert_err = ssl.set_der_cert(der_cert)
+    if not set_cert_ok then
+        return "failed to set DER cert: " .. set_cert_err
+    end
+
+    local der_priv_key, dev_priv_key_err = ssl.priv_key_pem_to_der(pem_cert_key)
+    if not der_priv_key then
+        return "failed to convert private key from PEM to DER: " .. dev_priv_key_err
+    end
+
+    local set_priv_key_ok, set_priv_key_err = ssl.set_der_priv_key(der_priv_key)
+    if not set_priv_key_ok then
+        return "failed to set DER private key: " .. set_priv_key_err
+    end
+end
+
+function _M.call()
+    local hostname = ssl.server_name()
+
+    local pem_cert_key = configuration.get_pem_cert_key(hostname)
+    if not pem_cert_key or pem_cert_key == "" then
+        ngx.log(ngx.ERR, "Certificate not found for the given hostname: " .. hostname)
```

Comment:
let's do `tostring(hostname)` to avoid errors when `hostname` is `nil`

Reply:
> pem_cert_key == ""

I'm not sure if we should handle this here or let `set_pem_cert_key` return an error. We should be doing input validation only once and that place is when we receive that data, in this case configuration module itself. Otherwise it gets dirty. You can read more about related concepts at https://en.wikipedia.org/wiki/Defensive_programming.